components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/JdbcTemplate.java

```diff
@@ -0,0 +1,377 @@
+                    .getErrorMessage() + query, e);
+        }
+    }
+
+    /**
+     * Executes the jdbc insert/update query.
+     *
+     * @param query       The SQL for insert/update.
+     * @param queryFilter Query filter to prepared statement parameter binding.
+     * @param bean        the Domain object to be inserted/updated.
+     * @param <T>         return type of the object.
+     */
+    public <T> int executeInsert(String query, QueryFilter queryFilter, T bean, boolean fetchInsertedId)
+            throws DataAccessException {
+
+        try (Connection connection = dataSource.getConnection()) {
+            int resultId;
+            if (fetchInsertedId) {
+                String dbProductName = connection.getMetaData().getDatabaseProductName();
+                try (PreparedStatement preparedStatement = connection.prepareStatement(query, new String[]{
+                        JdbcUtils.getConvertedAutoGeneratedColumnName(dbProductName, ID)})) {
+                    doInternalUpdate(queryFilter, preparedStatement);
+                    logDebug("Mapping generated key (Auto Increment ID) to the object");
+                    try (ResultSet generatedKeys = preparedStatement.getGeneratedKeys()) {
+                        if (generatedKeys.next()) {
+                            resultId = generatedKeys.getInt(1);
+                            logDebug("Newly inserted ID (Auto Increment ID) is {} for the bean {} ", resultId,
+                                    bean);
+                        } else {
+                            throw new SQLException(JdbcConstants.ErrorCodes.ERROR_CODE_AUTO_GENERATED_ID_FAILURE.
+                                    getErrorMessage());
+                        }
+                    }
+                }
+                if (!connection.getAutoCommit()) {
+                    connection.commit();
+                }
+                return resultId;
+            } else {
+                try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+                    doInternalUpdate(queryFilter, preparedStatement);
+                }
+                if (!connection.getAutoCommit()) {
+                    connection.commit();
+                }
+            }
+        } catch (SQLException e) {
+            logDebug("Error in performing database insert: {} with parameters {}", query, queryFilter);
+            throw new DataAccessException(JdbcConstants.ErrorCodes.ERROR_CODE_DATABASE_QUERY_PERFORMING.
+                    getErrorMessage() + query, JdbcConstants.ErrorCodes.ERROR_CODE_DATABASE_QUERY_PERFORMING.
+                    getErrorCode(), e);
+        }
+        return 0;
+    }
+
+    /**
+     * Executes the jdbc insert/update query.
+     *
+     * @param query       The SQL for insert/update.
+     * @param queryFilter Query filter to prepared statement parameter binding.
+     * @param bean        the Domain object to be inserted/updated.
```

Comment:
Shall we complete the method comment with the return type and throws description? Also, shall we fix this in other places?